src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -1103,28 +1150,83 @@ private ulong ToUInt64()
                 case CorElementType.ELEMENT_TYPE_CHAR:
                     return Unsafe.As<byte, ushort>(ref data);
                 case CorElementType.ELEMENT_TYPE_I4:
+#if TARGET_32BIT
+                case CorElementType.ELEMENT_TYPE_I:
+#endif
                     return (ulong)Unsafe.As<byte, int>(ref data);
                 case CorElementType.ELEMENT_TYPE_U4:
+#if TARGET_32BIT
+                case CorElementType.ELEMENT_TYPE_U:
+#endif
                 case CorElementType.ELEMENT_TYPE_R4:
                     return Unsafe.As<byte, uint>(ref data);
                 case CorElementType.ELEMENT_TYPE_I8:
+#if TARGET_64BIT
+                case CorElementType.ELEMENT_TYPE_I:
+#endif
                     return (ulong)Unsafe.As<byte, long>(ref data);
                 case CorElementType.ELEMENT_TYPE_U8:
+#if TARGET_64BIT
+                case CorElementType.ELEMENT_TYPE_U:
+#endif
                 case CorElementType.ELEMENT_TYPE_R8:
                     return Unsafe.As<byte, ulong>(ref data);
-                case CorElementType.ELEMENT_TYPE_I:
-                    return (ulong)Unsafe.As<byte, IntPtr>(ref data);
-                case CorElementType.ELEMENT_TYPE_U:
-                    return (ulong)Unsafe.As<byte, UIntPtr>(ref data);
                 default:
-                    throw new InvalidOperationException(SR.InvalidOperation_UnknownEnumType);
+                    Debug.Fail("Unknown enum underlying type");
+                    return 0;
             }
         }
 
         #endregion
 
         #region Object Overrides
 
+        public override bool Equals([NotNullWhen(true)] object? obj)
```

Comment:
Since Enum doesn't have `Equals<T>` I guess code like this([sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8BYAKDIAEBGJAAmAggBtaAhCBASwDsBzACgCyATwCi3AK75aMKmlojxUmQCYAlLQC8APhlUAdKICOE7E1z8Y6gNxk7pGJOmKntAN4BBNKzQBhAL5AA=) will always allocate 🙁. I guess if we could mark this one as AggressiveInlining and help jit to avoid box-unbox we would optimize it

Reply:
If we're going to invest in JIT support for enum, I'd prefer we do so by avoiding the boxing you get when doing "someEnum.ToString()", and then also adding an ISpanFormattable implementation that also doesn't box.

Reply:
btw, the jit opt I was referring is to handle something like this: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGue/uAhgBUAngAcYImLgzkAFAEsAdhgYIAlAwC8APmmzS8zQG4hQsZIMYjEYACsYYNRC16GEAHQBxGBksw8hpm1AC+QA==

Reply:
> the jit opt I was referring is to handle something like this

Would the existing opt-in escape analysis / stack allocation support catch this?

Reply:
I think so (but according to Andy it needs some more work). Here is a minimal repro:
```csharp
using System;
using System.Runtime.CompilerServices;

public class Program
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    static bool Test1(object? o1, object? o2)
    {
        if (o1 == null && o2 == null)
            return true;

        if (o1 == null || o2 == null)
            return false;

        if (o1.GetType() == o2.GetType())
            return GetRawData(o1) == GetRawData(o2);

        return false;
    }

    static ref byte GetRawData(object obj) =>
        ref Unsafe.As<RawData>(obj).Data;

    sealed class RawData
    {
        public byte Data;
    }

    public static void Main() => Console.WriteLine(Test1(42, 42));
}
```
Note: we already are able to fold that o1.GetType() != o2.GetType() on top of boxed objects

BTW, a quick fix for jit that catches GetType(BOX) in morph: https://github.com/EgorBo/runtime-1/commit/212c9d23a674d0c08dfa1ce3f3700fcb061df668 but the jit-diff is small because BOX is usually spilled to a temp

Reply:
>  I guess if we could mark this one as AggressiveInlining and help jit to avoid box-unbox we would optimize it

I would expect `Enum.Equals(object)` to be used very rarely. And when it is used, it is unlikely to be a path where performance matters.

> If we're going to invest in JIT support for enum, I'd prefer we do so by avoiding the boxing you get when doing "someEnum.ToString()", and then also adding an ISpanFormattable implementation that also doesn't box.

+1